# Data Quality Preparation
In this section I will be cleaning and preparing the vehicles and trip tables for the prediction model. It will be divided into two parts as they will be cleaned and prepared seperately.

#### Introduction:

### Trip Dataset
Each row represents one trip(route)
#### Understanding the features:
* DATASOURCE: Unique Bus Operator Code
* DAYOFSERVICE: Day of service. One day of service could last more than 24 hours
* TRIPID: Unique Trip code
* LINEID: Unique Line code
* ROUTEID: Unique route code
* DIRECTION: Route direction: (2)IB = inbound / going / northbound / eastbound, (1)OB = outbound / back / southbound / westbound
* PLANNEDTIME_ARR: Planned arrival time of the trip, in seconds
* PLANNEDTIME_DEP: Planned departure time of the trip, in seconds
* ACTUALTIME_ARR: Actual arrival time of the trip, in seconds
* ACTUALTIME_DEP: Actual departure time of the trip, in seconds
* BASIN: basin code
* TENDERLOT: tender lot
* SUPPRESSED: The whole trip has been supressed (0 = achieved, 1 = suppressed)
* JUSTIFICATIONID: Fault code
* LASTUPDATE: Time of the last record update 
* NOTE: Free note

### Vehicles Dataset
Each record of this table represents the service of one vehicle for one day of service and tells about
the overall distance and time worked by the associated vehicle in that specific day.

#### Understanding the features:
* DATASOURCE: Unique Bus Operator Code
* DAYOFSERVICE: Day of service
* VEHICLEID: Unique vehicle code arriving at this stop point
* DISTANCE: Distance travelled by the vehicle in the corresponding day
* MINUTES: Time worked by the veihcle in the corresponding day
* LASTUPDATE: Time of the last record update
* NOTE: Free note

In [1]:
# Importing the modules that might be used
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import dask.dataframe as dd
import sqlite3
from sqlalchemy import create_engine
from pprint import pprint

# ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Loading in the data
trips = pd.read_csv('/home/faye/data/rt_trips_DB_2018.txt', sep=';',error_bad_lines=False)
vehicles = pd.read_csv('/home/faye/data/rt_vehicles_DB_2018.txt', sep=';',error_bad_lines=False)

## Trips 
*Exploring the data*
<br><br>

In [3]:
trips.head(50)

DATASOURCE        DAYOFSERVICE   TRIPID LINEID  ROUTEID  DIRECTION  \
0          DB  07-FEB-18 00:00:00  6253783     68    68_80          1   
1          DB  07-FEB-18 00:00:00  6262138    25B  25B_271          2   
2          DB  07-FEB-18 00:00:00  6254942    45A   45A_70          2   
3          DB  07-FEB-18 00:00:00  6259460    25A  25A_273          1   
4          DB  07-FEB-18 00:00:00  6253175     14    14_15          1   
5          DB  07-FEB-18 00:00:00  6248240    77A   77A_28          2   
6          DB  07-FEB-18 00:00:00  6251760     39    39_21          2   
7          DB  07-FEB-18 00:00:00  6262909     16    16_20          1   
8          DB  07-FEB-18 00:00:00  6251147    40D  40D_102          1   
9          DB  07-FEB-18 00:00:00  6249435    27B   27B_34          2   
10         DB  07-FEB-18 00:00:00  6251483    142    142_9          2   
11         DB  07-FEB-18 00:00:00  6254310     83    83_22          2   
12         DB  07-FEB-18 00:00:00  6258441     14    14_15          1   
13         DB  07-FEB-18 00:00:00  6247622    130   130_10          1   
14         DB  07-FEB-18 00:00:00  6256532     15    15_16          1   
15         DB  07-FEB-18 00:00:00  6253924    46A   46A_74          1   
16         DB  07-FEB-18 00:00:00  6256349     33    33_44          1   
17         DB  07-FEB-18 00:00:00  6259785      7     7_70          2   
18         DB  07-FEB-18 00:00:00  6246028    39A   39A_40          1   
19         DB  07-FEB-18 00:00:00  6261203    39A   39A_43          2   
20         DB  07-FEB-18 00:00:00  6262164    46A   46A_74          1   
21         DB  07-FEB-18 00:00:00  6258154     49    49_28          2   
22         DB  07-FEB-18 00:00:00  6253554      1     1_40          2   
23         DB  07-FEB-18 00:00:00  6246745     15    15_16          1   
24         DB  07-FEB-18 00:00:00  6246014    123   123_34          1   
25         DB  07-FEB-18 00:00:00  6254029     83    83_22          2   
26         DB  07-FEB-18 00:00:00  6256059     41     41_7          2   
27         DB  07-FEB-18 00:00:00  6257123    67X   67X_38          1   
28         DB  07-FEB-18 00:00:00  6259641     59    59_10          1   
29         DB  07-FEB-18 00:00:00  6245836      9      9_5          1   
30         DB  07-FEB-18 00:00:00  6252377    130   130_10          1   
31         DB  07-FEB-18 00:00:00  6254347     40    40_27          1   
32         DB  07-FEB-18 00:00:00  6251013    39A   39A_40          1   
33         DB  07-FEB-18 00:00:00  6260186    239   239_28          2   
34         DB  07-FEB-18 00:00:00  6253738     76    76_37          2   
35         DB  07-FEB-18 00:00:00  6247623    130   130_11          2   
36         DB  07-FEB-18 00:00:00  6260764     84    84_27          1   
37         DB  07-FEB-18 00:00:00  6261193     53    53_20          1   
38         DB  18-FEB-18 00:00:00  6268949    185   185_55          1   
39         DB  18-FEB-18 00:00:00  6268951    185   185_55          1   
40         DB  18-FEB-18 00:00:00  6268953    185   185_55          1   
41         DB  18-FEB-18 00:00:00  6268387     16    16_20          1   
42         DB  18-FEB-18 00:00:00  6277453     16    16_20          1   
43         DB  18-FEB-18 00:00:00  6277810     16    16_20          1   
44         DB  18-FEB-18 00:00:00  6272110     41     41_3          1   
45         DB  18-FEB-18 00:00:00  6278258     41     41_3          1   
46         DB  18-FEB-18 00:00:00  6270605     41     41_3          1   
47         DB  18-FEB-18 00:00:00  6279865     41     41_3          1   
48         DB  18-FEB-18 00:00:00  6275294     41     41_3          1   
49         DB  18-FEB-18 00:00:00  6274325     41     41_3          1   

    PLANNEDTIME_ARR  PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP   BASIN  \
0             87245            84600         87524.0         84600.0  BasDef   
1             30517            26460         32752.0             NaN  BasDef   
2             35512            32100   

In [4]:
trips.tail(50)

DATASOURCE        DAYOFSERVICE   TRIPID LINEID  ROUTEID  DIRECTION  \
2182587         DB  14-MAY-18 00:00:00  6759988     13    13_60          1   
2182588         DB  14-MAY-18 00:00:00  6760080    140   140_19          1   
2182589         DB  14-MAY-18 00:00:00  6760086    111    111_8          2   
2182590         DB  14-MAY-18 00:00:00  6760100    123   123_34          1   
2182591         DB  14-MAY-18 00:00:00  6760154     76    76_37          2   
2182592         DB  14-MAY-18 00:00:00  6763776     33    33_72          2   
2182593         DB  14-MAY-18 00:00:00  6763777     33    33_45          1   
2182594         DB  14-MAY-18 00:00:00  6763797     84    84_31          2   
2182595         DB  14-MAY-18 00:00:00  6763806    151   151_15          1   
2182596         DB  14-MAY-18 00:00:00  6763831    151   151_16          1   
2182597         DB  14-MAY-18 00:00:00  6763881    17A   17A_17          2   
2182598         DB  14-MAY-18 00:00:00  6757370     67     67_6          2   
2182599         DB  14-MAY-18 00:00:00  6757400     47   47_139          1   
2182600         DB  14-MAY-18 00:00:00  6757403     47   47_135          2   
2182601         DB  14-MAY-18 00:00:00  6757455    161   161_51          2   
2182602         DB  14-MAY-18 00:00:00  6757468     33    33_44          1   
2182603         DB  14-MAY-18 00:00:00  6757509    27A    27A_4          1   
2182604         DB  14-MAY-18 00:00:00  6757514    27A    27A_5          2   
2182605         DB  14-MAY-18 00:00:00  6757515    27A    27A_4          1   
2182606         DB  14-MAY-18 00:00:00  6762371    130   130_11          2   
2182607         DB  14-MAY-18 00:00:00  6762374    130   130_10          1   
2182608         DB  14-MAY-18 00:00:00  6762495    25A  25A_270          2   
2182609         DB  14-MAY-18 00:00:00  6762520    45A   45A_60          1   
2182610         DB  14-MAY-18 00:00:00  6763184     53    53_20          1   
2182611         DB  14-MAY-18 00:00:00  6762805    140   140_21          2   
2182612         DB  14-MAY-18 00:00:00  6762895     27    27_19          1   
2182613         DB  14-MAY-18 00:00:00  6762930      9      9_7          2   
2182614         DB  14-MAY-18 00:00:00  6762969     18     18_3          1   
2182615         DB  14-MAY-18 00:00:00  6764976     7A    7A_87          2   
2182616         DB  14-MAY-18 00:00:00  6765053    54A   54A_12          2   
2182617         DB  14-MAY-18 00:00:00  6765067     26    26_24          1   
2182618         DB  14-MAY-18 00:00:00  6765072    56A   56A_31          1   
2182619         DB  14-MAY-18 00:00:00  6765081    56A   56A_30          2   
2182620         DB  14-MAY-18 00:00:00  6765101    79A   79A_28          2   
2182621         DB  14-MAY-18 00:00:00  6756423    40D  40D_102          1   
2182622         DB  14-MAY-18 00:00:00  6756429    40D  40D_102          1   
2182623         DB  14-MAY-18 00:00:00  6767663    40D  40D_104          2   
2182624         DB  14-MAY-18 00:00:00  6754588     41     41_7          2   
2182625         DB  14-MAY-18 00:00:00  6757965    151   151_15          1   
2182626         DB  14-MAY-18 00:00:00  6758027    145  145_105          2   
2182627         DB  14-MAY-18 00:00:00  6765635    130   130_11          2   
2182628         DB  14-MAY-18 00:00:00  6765649    130   130_11          2   
2182629         DB  14-MAY-18 00:00:00  6765652     53    53_21          2   
2182630         DB  14-MAY-18 00:00:00  6765662     53    53_21          2   
2182631         DB  14-MAY-18 00:00:00  6765828    45A   45A_68          1   
2182632         DB  14-MAY-18 00:00:00  6765849    123   123_36          2   
2182633         DB  14-MAY-18 00:00:00  6765469     75    75_17          1   
2182634         DB  14-MAY-18 00:00:00  6765486    33D   33D_62          2   
2182635         DB  14-MAY-18 00:00:00  6764987     70    70_60          1   
2182636         DB  14-MAY-18 00:00:00  6765012     27    27_19          1   

         PLANNEDTIME_ARR  PLA

In [5]:
# Columns and rows
trips.shape

(2182637, 16)

In [6]:
trips.dtypes

DATASOURCE          object
DAYOFSERVICE        object
TRIPID               int64
LINEID              object
ROUTEID             object
DIRECTION            int64
PLANNEDTIME_ARR      int64
PLANNEDTIME_DEP      int64
ACTUALTIME_ARR     float64
ACTUALTIME_DEP     float64
BASIN               object
TENDERLOT          float64
SUPPRESSED         float64
JUSTIFICATIONID    float64
LASTUPDATE          object
NOTE                object
dtype: object

In [7]:
trips.isnull().sum()

DATASOURCE               0
DAYOFSERVICE             0
TRIPID                   0
LINEID                   0
ROUTEID                  0
DIRECTION                0
PLANNEDTIME_ARR          0
PLANNEDTIME_DEP          0
ACTUALTIME_ARR      137207
ACTUALTIME_DEP      164551
BASIN                    0
TENDERLOT          2182637
SUPPRESSED         2178304
JUSTIFICATIONID    2178307
LASTUPDATE               0
NOTE                     0
dtype: int64

In [8]:
trips.nunique()

DATASOURCE              1
DAYOFSERVICE          360
TRIPID             658964
LINEID                130
ROUTEID               588
DIRECTION               2
PLANNEDTIME_ARR     64461
PLANNEDTIME_DEP       791
ACTUALTIME_ARR      68122
ACTUALTIME_DEP      66771
BASIN                   1
TENDERLOT               0
SUPPRESSED              1
JUSTIFICATIONID      3526
LASTUPDATE            360
NOTE                46690
dtype: int64

In [9]:
# trips_columns = list(trips.columns.values)

In [10]:
# Seperating features
categorical_columns = ['DATASOURCE','TRIPID','LINEID','ROUTEID','DIRECTION','BASIN','TENDERLOT','JUSTIFICATIONID','NOTE','SUPPRESSED']
time_columns = ['PLANNEDTIME_ARR','PLANNEDTIME_DEP','ACTUALTIME_ARR','ACTUALTIME_DEP']
datetime = ['DAYOFSERVICE','LASTUPDATE']

# Converting seconds into times and into datetime series
import datetime
for row in trips['PLANNEDTIME_ARR']:
    trips['PLANNEDTIME_ARR'] = trips['PLANNEDTIME_ARR'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['PLANNEDTIME_DEP']:
    trips['PLANNEDTIME_DEP'] = trips['PLANNEDTIME_DEP'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['ACTUALTIME_ARR']:
    trips['ACTUALTIME_ARR'] = trips['ACTUALTIME_ARR'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['ACTUALTIME_DEP']:
    trips['ACTUALTIME_DEP'] = trips['ACTUALTIME_DEP'].replace(row, str(datetime.timedelta(seconds=row)))

In [11]:
# Converting objects into categorical types
for column in categorical_columns:
    trips[column] = trips[column].astype('category')

# Converting objects into datetime series
for column2 in datetime:
    trips[column2] = pd.to_datetime(trips[column2])

In [12]:
trips.dtypes

DATASOURCE               category
DAYOFSERVICE       datetime64[ns]
TRIPID                   category
LINEID                   category
ROUTEID                  category
DIRECTION                category
PLANNEDTIME_ARR             int64
PLANNEDTIME_DEP             int64
ACTUALTIME_ARR            float64
ACTUALTIME_DEP            float64
BASIN                    category
TENDERLOT                category
SUPPRESSED               category
JUSTIFICATIONID          category
LASTUPDATE         datetime64[ns]
NOTE                     category
dtype: object

In [13]:
# Converting the time columns seconds into times. The days part should be ignored.
for column3 in time_columns:
    trips[column3] = trips[column3].apply(pd.to_timedelta, unit='s')
    trips[column3] = trips[column3]- pd.to_timedelta(trips[column3].dt.days, unit='d')

In [14]:
trips.head(10)

DATASOURCE DAYOFSERVICE   TRIPID LINEID  ROUTEID DIRECTION PLANNEDTIME_ARR  \
0         DB   2018-02-07  6253783     68    68_80         1 0 days 00:14:05   
1         DB   2018-02-07  6262138    25B  25B_271         2 0 days 08:28:37   
2         DB   2018-02-07  6254942    45A   45A_70         2 0 days 09:51:52   
3         DB   2018-02-07  6259460    25A  25A_273         1 0 days 15:54:21   
4         DB   2018-02-07  6253175     14    14_15         1 0 days 23:43:03   
5         DB   2018-02-07  6248240    77A   77A_28         2 0 days 11:34:08   
6         DB   2018-02-07  6251760     39    39_21         2 0 days 09:39:28   
7         DB   2018-02-07  6262909     16    16_20         1 0 days 12:12:16   
8         DB   2018-02-07  6251147    40D  40D_102         1 0 days 09:19:18   
9         DB   2018-02-07  6249435    27B   27B_34         2 0 days 14:38:02   

  PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP   BASIN TENDERLOT  \
0 0 days 23:30:00 0 days 00:18:44 0 days 23:30:00  BasDef       NaN   
1 0 days 07:21:00 0 days 09:05:52             NaT  BasDef       NaN   
2 0 days 08:55:00 0 days 10:05:29 0 days 08:54:42  BasDef       NaN   
3 0 days 15:07:00 0 days 16:14:23 0 days 15:07:23  BasDef       NaN   
4 0 days 22:40:00 0 days 23:31:22 0 days 22:40:08  BasDef       NaN   
5 0 days 10:20:00 0 days 11:40:19 0 days 10:25:38  BasDef       NaN   
6 0 days 08:02:00 0 days 09:55:09 0 days 08:02:09  BasDef       NaN   
7 0 days 10:48:00 0 days 12:08:41 0 days 10:49:15  BasDef       NaN   
8 0 days 08:30:00 0 days 09:35:40 0 days 08:39:45  BasDef       NaN   
9 0 days 13:50:00 0 days 14:48:01 0 days 13:52:54  BasDef       NaN   

  SUPPRESSED JUSTIFICATIONID          LASTUPDATE       NOTE  
0        NaN             NaN 2018-02-28 12:05:11  ,2967409,  
1        NaN             NaN 2018-02-28 12:05:11  ,2580260,  
2        NaN             NaN 2018-02-28 12:05:11  ,2448968,  
3        NaN             NaN 2018-02-28 12:05:11  ,3094242,  
4        NaN             NaN 2018-02-28 12:05:11  ,2526331,  
5        NaN             NaN 2018-02-28 12:05:11  ,2966500,  
6        NaN             NaN 2018-02-28 12:05:11  ,2422850,  
7        NaN             NaN 2018-02-28 12:05:11  ,3045261,  
8        NaN             NaN 2018-02-28 12:05:11  ,2837123,  
9        NaN             NaN 2018-02-28 12:05:11  ,2329338,

In [22]:
trips_sorted = trips.sort_values(by='TRIPID')
trips_sorted.head(50)

DATASOURCE DAYOFSERVICE   TRIPID LINEID ROUTEID DIRECTION  \
681660          DB   2018-01-06  5955221     14   14_15         1   
723197          DB   2018-01-06  5955222     14   14_16         2   
684398          DB   2018-01-06  5955223     14   14_15         1   
719581          DB   2018-01-06  5955224     14   14_16         2   
674864          DB   2018-01-06  5955225     14   14_15         1   
2064218         DB   2018-01-03  5955226    77A  77A_29         1   
2158096         DB   2018-01-02  5955226    77A  77A_29         1   
231424          DB   2018-01-05  5955226    77A  77A_29         1   
737493          DB   2018-01-04  5955226    77A  77A_29         1   
225052          DB   2018-01-05  5955227    77A  77A_28         2   
2057152         DB   2018-01-03  5955227    77A  77A_28         2   
2155489         DB   2018-01-02  5955227    77A  77A_28         2   
2119594         DB   2018-01-04  5955227    77A  77A_28         2   
732682          DB   2018-01-04  5955228    77A  77A_29         1   
242014          DB   2018-01-05  5955228    77A  77A_29         1   
2061908         DB   2018-01-03  5955228    77A  77A_29         1   
2161694         DB   2018-01-02  5955228    77A  77A_29         1   
2101746         DB   2018-01-04  5955229    77A  77A_28         2   
674979          DB   2018-01-05  5955229    77A  77A_28         2   
749640          DB   2018-01-02  5955229    77A  77A_28         2   
694994          DB   2018-01-03  5955229    77A  77A_28         2   
2057153         DB   2018-01-03  5955230    77A  77A_29         1   
2161612         DB   2018-01-02  5955230    77A  77A_29         1   
2119595         DB   2018-01-04  5955230    77A  77A_29         1   
240737          DB   2018-01-05  5955230    77A  77A_29         1   
2068181         DB   2018-01-03  5955231    77A  77A_28         2   
2154036         DB   2018-01-02  5955231    77A  77A_28         2   
745588          DB   2018-01-04  5955231    77A  77A_28         2   
222252          DB   2018-01-05  5955231    77A  77A_28         2   
2152788         DB   2018-01-02  5955232    77A  77A_29         1   
215586          DB   2018-01-05  5955232    77A  77A_29         1   
214086          DB   2018-01-06  5955233     25  25_269         2   
214075          DB   2018-01-06  5955234     25  25_272         1   
214076          DB   2018-01-06  5955235     25  25_269         2   
211744          DB   2018-01-06  5955236     25  25_272         1   
715388          DB   2018-01-06  5955237     25  25_269         2   
719593          DB   2018-01-06  5955238     25  25_272         1   
720912          DB   2018-01-06  5955239     25  25_269         2   
661290          DB   2018-01-06  5955240     25  25_272         1   
723198          DB   2018-01-06  5955241     25  25_269         2   
719594          DB   2018-01-06  5955242     25  25_272         1   
681661          DB   2018-01-06  5955243     25  25_269         2   
719477          DB   2018-01-06  5955244     25  25_272         1   
680442          DB   2018-01-06  5955245     25  25_269         2   
719595          DB   2018-01-06  5955246     25  25_272         1   
723199          DB   2018-01-06  5955247     25  25_269         2   
214087          DB   2018-01-06  5955248     25  25_272         1   
715389          DB   2018-01-06  5955249     25  25_269         2   
211745          DB   2018-01-06  5955250    46A  46A_67         2   
652225          DB   2018-01-06  5955251    46A  46A_74         1   

        PLANNEDTIME_ARR PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP  \
681660  0 days 11:49:13 0 days 10:30:00 0 days 11:41:18 0 days 10:30:36   
723197  0 days 13:17:47 0 days 12:00:00 0 days 13:15:07 0 days 11:59:42   
684398  0 days 15:01:45 0 days 13:40:00 0 days 15:02:08 0 days 13:39:31   
719581  0 days 16:41:34 0 days 15:20:00 0 days 16:40:25 0 days 15:20:03   
674864  0 days 18:15:51 0 days 17:00:00 0 days 18:08:52 0 days 16:59:21   
2064218 0 days 08:18:00 0 days 07:20:0

In [34]:
trip77a = trips_sorted.loc[trips_sorted['LINEID']=='77A']
trip77a = trip77a.loc[trips_sorted['DAYOFSERVICE']=='2018-01-03']
trip77a = trip77a.sort_values(by='PLANNEDTIME_ARR')
trip77a.head(50)

DATASOURCE DAYOFSERVICE   TRIPID LINEID ROUTEID DIRECTION  \
694969          DB   2018-01-03  5965215    77A  77A_29         1   
2059759         DB   2018-01-03  5969376    77A  77A_28         2   
692364          DB   2018-01-03  5964114    77A  77A_29         1   
707213          DB   2018-01-03  5957107    77A  77A_29         1   
714026          DB   2018-01-03  5959625    77A  77A_29         1   
703209          DB   2018-01-03  5956817    77A  77A_29         1   
2062834         DB   2018-01-03  5964103    77A  77A_28         2   
697807          DB   2018-01-03  5961549    77A  77A_29         1   
2049037         DB   2018-01-03  5971472    77A  77A_28         2   
701813          DB   2018-01-03  5957040    77A  77A_29         1   
689488          DB   2018-01-03  5969367    77A  77A_28         2   
2064218         DB   2018-01-03  5955226    77A  77A_29         1   
709963          DB   2018-01-03  5957108    77A  77A_28         2   
714014          DB   2018-01-03  5964104    77A  77A_29         1   
693708          DB   2018-01-03  5959626    77A  77A_28         2   
2047686         DB   2018-01-03  5970605    77A  77A_28         2   
711373          DB   2018-01-03  5956818    77A  77A_28         2   
2064908         DB   2018-01-03  5971473    77A  77A_29         1   
692247          DB   2018-01-03  5970294    77A  77A_28         2   
2057252         DB   2018-01-03  5957220    77A  77A_28         2   
703215          DB   2018-01-03  5969368    77A  77A_29         1   
693716          DB   2018-01-03  5961550    77A  77A_28         2   
2054455         DB   2018-01-03  5965206    77A  77A_28         2   
707214          DB   2018-01-03  5957109    77A  77A_29         1   
2062781         DB   2018-01-03  5957041    77A  77A_28         2   
2047683         DB   2018-01-03  5971830    77A  77A_28         2   
2047692         DB   2018-01-03  5959627    77A  77A_29         1   
2057152         DB   2018-01-03  5955227    77A  77A_28         2   
697809          DB   2018-01-03  5970606    77A  77A_29         1   
714015          DB   2018-01-03  5964105    77A  77A_28         2   
687084          DB   2018-01-03  5970295    77A  77A_29         1   
2064909         DB   2018-01-03  5971474    77A  77A_28         2   
714056          DB   2018-01-03  5969369    77A  77A_28         2   
2068952         DB   2018-01-03  5965207    77A  77A_29         1   
2054489         DB   2018-01-03  5957110    77A  77A_28         2   
692265          DB   2018-01-03  5971831    77A  77A_29         1   
704510          DB   2018-01-03  5959628    77A  77A_28         2   
2061908         DB   2018-01-03  5955228    77A  77A_29         1   
687041          DB   2018-01-03  5970607    77A  77A_28         2   
2067264         DB   2018-01-03  5964106    77A  77A_29         1   
696429          DB   2018-01-03  5970296    77A  77A_28         2   
2047645         DB   2018-01-03  5971475    77A  77A_29         1   
2062884         DB   2018-01-03  5965208    77A  77A_28         2   
703216          DB   2018-01-03  5969370    77A  77A_29         1   
2065540         DB   2018-01-03  5957111    77A  77A_29         1   
694994          DB   2018-01-03  5955229    77A  77A_28         2   
712731          DB   2018-01-03  5959629    77A  77A_29         1   
712654          DB   2018-01-03  5964107    77A  77A_28         2   
703265          DB   2018-01-03  5970608    77A  77A_29         1   
2064855         DB   2018-01-03  5970297    77A  77A_29         1   

        PLANNEDTIME_ARR PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP  \
694969  0 days 00:04:24 0 days 23:00:00             NaT 0 days 22:58:27   
2059759 0 days 00:25:09 0 days 23:30:00 0 days 00:14:56 0 days 23:29:40   
692364  0 days 00:29:24 0 days 23:25:00             NaT 0 days 23:25:04   
707213  0 days 06:38:00 0 days 05:40:00 0 days 06:33:55 0 days 05:44:10   
714026  0 days 06:58:00 0 days 06:00:00 0 days 06:54:40 0 days 05:59:25   
703209  0 days 07:18:00 0 days 06:20:0

## Note about this specific route
In this route, we took the route 77a from a date (January 3, 2018) and sorted it by PLANNEDTIME_ARR. It has a total of 101 rows. The first few values of PLANNEDTIME_ARR, these may be the last few trips that ran that day and it was just sorted as the top? In any case, We see that they provide data for each route for one day with all of the times. It is a matter of matching it with the leavetimes dataset then. This will be explored at the very end. 

## Duplicates
Exploring if the trips dataset has any duplicates. There shouldn't be duplicates for any reason. The trips only siginify one route.

We see there aren't any duplicates so this is a good sign for the data. 

In [17]:
duplicates = trips[trips.duplicated(keep=False)]
print("Number of rows that are duplicates: ", duplicates.shape[0])

Number of rows that are duplicates:  0


## Checking descriptive columns

In [18]:
trips[datetime].describe(datetime_is_numeric=True).T

count                          mean                 min  \
DAYOFSERVICE  2182637 2018-06-28 17:25:38.969786880 2018-01-01 00:00:00   
LASTUPDATE    2182637 2018-07-16 10:25:44.996667392 2018-01-08 17:21:10   

                             25%                 50%                 75%  \
DAYOFSERVICE 2018-03-31 00:00:00 2018-06-28 00:00:00 2018-09-26 00:00:00   
LASTUPDATE   2018-04-10 15:08:12 2018-07-13 19:47:00 2018-10-23 06:57:21   

                             max  
DAYOFSERVICE 2018-12-31 00:00:00  
LASTUPDATE   2019-01-16 18:27:21

In [19]:
trips[time_columns].describe().T

count                      mean                       std  \
PLANNEDTIME_ARR  2182637 0 days 14:58:33.083917298 0 days 05:01:58.847667725   
PLANNEDTIME_DEP  2182637 0 days 14:18:14.341037927 0 days 04:53:08.417924110   
ACTUALTIME_ARR   2045430 0 days 15:00:19.765848745 0 days 04:59:23.886687638   
ACTUALTIME_DEP   2018086 0 days 14:21:03.389364972 0 days 04:52:25.239888791   

                            min                    25%             50%  \
PLANNEDTIME_ARR 0 days 00:00:00        0 days 10:47:14 0 days 15:08:10   
PLANNEDTIME_DEP 0 days 00:05:00        0 days 10:00:00 0 days 14:20:00   
ACTUALTIME_ARR  0 days 00:00:00        0 days 10:48:42 0 days 15:07:59   
ACTUALTIME_DEP  0 days 00:00:07 0 days 10:07:17.250000 0 days 14:20:09   

                            75%             max  
PLANNEDTIME_ARR 0 days 18:58:40 0 days 23:59:59  
PLANNEDTIME_DEP 0 days 18:05:00 0 days 23:50:00  
ACTUALTIME_ARR  0 days 19:00:00 0 days 23:59:59  
ACTUALTIME_DEP  0 days 18:09:41 0 days 23:59:58

In [20]:
trips[categorical_columns].describe().T

count  unique        top     freq
DATASOURCE       2182637       1         DB  2182637
TRIPID           2182637  658964    7324685       19
LINEID           2182637     130        46A    76728
ROUTEID          2182637     588     46A_74    37182
DIRECTION        2182637       2          2  1100273
BASIN            2182637       1     BasDef  2182637
TENDERLOT              0       0        NaN      NaN
JUSTIFICATIONID     4330    3526     203708       19
NOTE             2182637   46690  ,2326480,      244
SUPPRESSED          4333       1          0     4333

We see there a couple of features that have <= 1 cardinalities:
* DATASOURCE has 1 unique value.
* TRIPID has more than 600k unique values.
* LINEID has 130 unique values.
* ROUTEID has 588 unique values.
* DIRECTION has 2 unique values.
* BASIN has 1 unique values.
* TENDERLOT has no unique values meaning that it is just null.
* JUSTIFICATIONID has 4330 unique values.
* NOTE has more than 46k unique values.
* SUPPRESSED has 4333 unique values.

**RESULT :** As we can see, DATASOURCE, BASIN, TENDERLOT and SUPPRESSED has 1 or 0 unique values. This means the same value is consistent throughout the dataset. Therefore, these will be dropped.

## Logical Integrity Tests

List of integrity tests to be performed:
* Make sure LASTUPDATE dates doesn't occur before DAYSERVICE
* Ensure that the timetables match the specific route and their PLANNEDTIME_DEP
* Ensure that the routeID has the matching route number in LINEID

#### TEST 1: Ensure LASTUPDATE doesn't occur before DAYOFSERVICE

In [40]:
test1 = trips[trips['DAYOFSERVICE'] > trips['LASTUPDATE']]
print("The number of rows that have dates that occur before DAYOFSERVICE from LASTUPDATES is: ", test1.shape[0])

The number of rows that have dates that occur before DAYOFSERVICE from LASTUPDATES is:  0


#### TEST 2: Ensure timetables match the specific route with PLANNEDTIME_DEP
This is to check that the timetables match with all of the routes. We take a look at PLANNEDTIME_DEP to check this. 

In [46]:
# Need to get a list of all of the routes and their timetables, then compare it with the routes 
# and their planned departure times

# Make a dataframe with just LINEID and PLANNEDTIME_DEP and make a list of all of the routes, order them
data = {'LINEID': trips['LINEID'], 'PLANNEDTIME_DEP': trips['PLANNEDTIME_DEP']}
line_dep = pd.DataFrame(data)
list_routes = list(line_dep.LINEID.unique())
list_routes = sorted(list_routes)


In [73]:
!pip3 install beautifulsoup4

  Using cached https://files.pythonhosted.org/packages/d1/41/e6495bd7d3781cee623ce23ea6ac73282a373088fcd0ddc809a047b18eae/beautifulsoup4-4.9.3-py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/36/69/d82d04022f02733bf9a72bc3b96332d360c0c5307096d76f6bb7489f7e57/soupsieve-2.2.1-py3-none-any.whl


* Connect to the dublinbus website
* Get the url for the one that shows the timetable for one route
* Loop through list_route and apply it at the end of the url
* Get the following:
    * times
    * day
* day search by 'strong'
* times search by 'div class="time"'
* Store the times values in a dataframe:
    * Weekday (monday - friday)
    * Saturday
    * Sunday

In [127]:
# Make a dataframe that scrapes the route numbers and the times within that route
import requests

# Making a timetables variable to store the link
url = 'https://www.dublinbus.ie/Your-Journey1/Timetables/All-Timetables/25/'
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')
find = soup.find(id='ctl00_FullRegion_MainRegion_MainContentRegion_MainBodyRegion_Timetable1_ctl02_DataList1_ctl00_dlTimeTop')

# Need to find a way to only get the interger and put it on a list
times_weekday = soup.find_all("div", {"class": "time"})

In [147]:
# Lets convert times_weekday into a list and then convert its values into a string
times_weekday = list(times_weekday)
for value in range(len(times_weekday)):
    times_weekday[value] = str(times_weekday[value])

# Then, using regex, get the interger and convert it to a string version of timedelta then put it in a list
import re
weekdays = []
for value in range(len(times_weekday)):
    parse = re.findall('\d+', times_weekday[value])
    time = str('0 days ' + parse[0] + ':' + parse[1] + ':00')
    weekdays += [time]

In [153]:
# These are not all weekdays
weekdays

['0 days 08:20:00',
 '0 days 09:20:00',
 '0 days 10:20:00',
 '0 days 11:20:00',
 '0 days 12:20:00',
 '0 days 13:20:00',
 '0 days 14:20:00',
 '0 days 15:20:00',
 '0 days 16:20:00',
 '0 days 16:50:00',
 '0 days 17:20:00',
 '0 days 18:30:00',
 '0 days 19:30:00',
 '0 days 20:30:00',
 '0 days 21:30:00',
 '0 days 22:30:00',
 '0 days 23:20:00',
 '0 days 07:05:00',
 '0 days 08:05:00',
 '0 days 09:05:00',
 '0 days 10:05:00',
 '0 days 11:05:00',
 '0 days 12:05:00',
 '0 days 13:05:00',
 '0 days 14:05:00',
 '0 days 15:05:00',
 '0 days 16:05:00',
 '0 days 17:05:00',
 '0 days 18:05:00',
 '0 days 19:05:00',
 '0 days 20:05:00',
 '0 days 21:05:00',
 '0 days 22:05:00',
 '0 days 23:05:00',
 '0 days 11:40:00',
 '0 days 13:40:00',
 '0 days 15:40:00',
 '0 days 17:40:00',
 '0 days 19:40:00',
 '0 days 21:40:00',
 '0 days 23:25:00',
 '0 days 06:40:00',
 '0 days 06:55:00',
 '0 days 07:10:00',
 '0 days 08:05:00',
 '0 days 09:20:00',
 '0 days 10:20:00',
 '0 days 11:20:00',
 '0 days 12:20:00',
 '0 days 13:20:00',


#### Converting seconds into times and into datetime series
import datetime
for row in trips['PLANNEDTIME_ARR']:
    trips['PLANNEDTIME_ARR'] = trips['PLANNEDTIME_ARR'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['PLANNEDTIME_DEP']:
    trips['PLANNEDTIME_DEP'] = trips['PLANNEDTIME_DEP'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['ACTUALTIME_ARR']:
    trips['ACTUALTIME_ARR'] = trips['ACTUALTIME_ARR'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['ACTUALTIME_DEP']:
    trips['ACTUALTIME_DEP'] = trips['ACTUALTIME_DEP'].replace(row, str(datetime.timedelta(seconds=row)))
    
trips1 = trips
trips2 = trips1['PLANNEDTIME_ARR']
trips2 = trips2.apply(pd.to_timedelta, unit='s')
trips2 = trips2 - pd.to_timedelta(trips2.dt.days, unit='d')
trips2.head()